Customize premixing stage1 and parts of stage2 with modifiers, and extend Modifier boolean expressions to include "not" and "or"

makortel opened the pull request:
Currently premixing customizations are implemented as separate files that import or copy from the standard ones, and get used in the configurations according to the cmsDriver parameters. This PR replaces the customizations in stage1 fully, and in stage2 partially, with "process modifier". This is a preparatory step for phase2 premixing developments (making phase2 customizations a bit easier to do). This PR also addresses #21704.

In a bit more detail this PR
* adds modifiers `premix_stage1` and `premix_stage2`
   * `--procModifiers` parameter is added to all premixing workflows
* removes `DIGIPREMIX` and `DIGIPREMIX_S2` steps from cmsDriver (both can be now just `DIGI`)
* migrates (i.e. removes) configuration files
   * stage1: `Digi_Premix_cff`, `DigiToRawPremixing_cff`, `SimL1EmulatorPremix_cff`, `digi_noNoise_cfi` and "`DIGIPREMIX` terrible hacks"
   * stage2: `digi_MixPreMix_cfi`, `DigiDMPremix_cff`

What remains for stage2 is at least `DataMixerPreMix_cff`. It wasn't clear to me whether moving the customizations to the default `DataMixerDataOnSim_cff` would actually be an improvement, so I decided to leave that for a later exercise (and then one would have to think the relationship of `--procModifiers premix_stage2` and `--datamix` cmsDriver parameters).

The idea has been discussed in #21704 and in simulation meeting on Feb 6th.
https://indico.cern.ch/event/702324/contributions/2885358/attachments/1596021/2527902/slides_mk_simulation_20180206.pdf

As a sidestep, in order to avoid adding `isChosen()` calls, the `Modifier` boolean expressions are extended with "not" (`~`)  and "or" (`|`). For completeness added support for `toReplaceWith()` to the resulting `Modifier` of the boolean expressions.

Tested in CMSSW_10_1_X_2018-02-11-2300, no changes expected in results. I've tested with workflows 250199.{0,17,18} and 250202.{0,17,18} (such that the 250202.X use premixing library from 250202.X as an input) that the DQM histograms agree, and that the configuration dumps (from both stage1 and stage2) do not contain essential differences (there are many changes visible not affecting any functionality because the "customization point" is moved earlier in the chain of cloning of PSets).

@mdhildreth @kpedro88

makortel committed:
Add premix_stage1 and premix_stage2 process modifiers

makortel committed:
Replace premixing stage1 Digi_PreMix_cff with modifier customization

_enableDigiAliases hack is no longer needed

Remove Digi_PreMix_cff as unnecessary

makortel committed:
Replace premixing stage1 DigiToRawPremixing_cff with modifier customization

Note that HCAL phase2 packing+unpacking was enabled in #20920, so
there is no need to remove that for phase2 premixing (which anyway is
not being tested or working at the moment).

Also note that in L1TDigiToRaw_cff we have to reset the
rawDataCollector.RawCollectionList as if the L1TDigiToRaw_cff would
not have been loaded. This was not needed earlier because the file was
explicitly not loaded.

makortel committed:
Replace premixing stage1 DIGIPREMIX "terrible hacks" with modifier customization

makortel committed:
Replace premixing stage1 SimL1EmulatorPremix_cff with modifier customization

makortel committed:
Replace premixing stage1 digi_noNoise_cfi with modifier customization

makortel committed:
Remove DIGIPREMIX as unnecessary

Fully replaced with premix_stage1 modifier

makortel committed:
Replace premixing stage2 digi_MixPreMix_cfi with modifier customization

makortel committed:
Remove DIGIPREMIX_S2 as unnecessary

makortel committed:
Replace premixing stage2 DigiDMPremix_cff with modifier customization

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22210/3357

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

Configuration/Applications
Configuration/ProcessModifiers
Configuration/PyReleaseValidation
Configuration/StandardSequences
EventFilter/ESDigiToRaw
EventFilter/SiStripRawToDigi
FastSimulation/Configuration
L1Trigger/Configuration
SimCalorimetry/Configuration
SimCalorimetry/EcalSelectiveReadoutProducers
SimCalorimetry/EcalSimProducers
SimCalorimetry/HcalSimProducers
SimCalorimetry/HcalZeroSuppressionProducers
SimGeneral/MixingModule
SimGeneral/PileupInformation
SimMuon/RPCDigitizer
SimTracker/SiPhase2Digitizer


@perrotta, @cmsbuild, @prebello, @lveldere, @emeschi, @civanch, @mdhildreth, @fabozzi, @nsmith-, @rekovic, @franzoni, @kpedro88, @thomreis, @slava77, @mommsen, @GurpreetSinghChahal, @fabiocos, @ssekmen, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @kreczko, @felicepantaleo, @sdevissc, @GiacomoSguazzoni, @jhgoh, @VinInn, @Martin-Grunewald, @ebrondol, @rovere, @matt-komm, @mariadalfonso, @argiro, @threus, @dgulhan this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test with 250199.0,250199.17,250199.18,250202.0,250202.17,250202.18

(even if these tests do not provide any evidence that the stage1 workflows 250199.X would still produce the same output)

davidlange6 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 087dced467bffa78834c3a817e0cc9ff3116977e
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26065/summary.html

cmsbuild commented:
Comparison job queued.

kpedro88 commented on Configuration/StandardSequences/python/Digi_cff.py:
```diff
@@ -42,6 +42,13 @@
 def _fastSimDigis(process):
+    # no need for the aliases for usual mixing
+    from Configuration.ProcessModifiers.premix_stage1_cff import premix_stage1
+    if premix_stage1.isChosen():
```
@Dr15Jones could we extend the modifier classes somehow to allow `toModify` or `toReplaceWith` to act on ProcessModifier objects? Then one could replace the `__func` assigned to it with a dummy function.

davidlange6 replied:
or one could also have multiple fastsim modifiers (fastsim, fastsim_premix1, ...) that then map more cleaning onto the existing interface

kpedro88 replied:
In this case I think it would have to be something like fastsim and fastsim_notpremix1 (since we want to omit this process modifier in premix_stage1), which is a bit counterintuitive.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26065/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/1325.7_TTbar_13_94XNanoAODINPUT+TTbar_13_94XNanoAODINPUT+NANOEDMMC2017+HARVESTNANOAODMC2017

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465190
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.27 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

Dr15Jones commented:
I'm with @davidlange6 on multiple modifiers.

makortel commented:
> I'm with @davidlange6 on multiple modifiers.

I can make a try to see how `fastsim_not_premix_stage1` would look like. A nasty feature is that it has to be put for all FastSim workflows except the premixing stage1 (and I don't know how much people generate FastSim configurations with cmsDriver directly, I vaguely recall seeing some twiki pages with cmsDriver recipes).

Is the other suggestion in the code comment, i.e. "_InverseModifier" out of the question then? With that the `makeProcessModifier()` line would look like
```python
modifyDigi_fastSimDigis = (fastSim & ~premix_stage1).makeProcessModifier(_fastSimDigis)
```

kpedro88 commented:
> A nasty feature is that it has to be put for all FastSim workflows except the premixing stage1

Yes, this seems like a dramatic and likely dangerous refactoring for a single change.

Dr15Jones commented:
I'm OK with extending the system to allow `~premix_stage1`.

makortel committed:
Add _InvertModifier and _OrModifier to allow more complex boolean expressions of Modifiers

For these to work I had to bypass the isChosen() checks of the
contained Modifiers as they may not be enabled.

makortel committed:
Do the sanity checks on Modifier.toModify() arguments upfront also for _{And,Invert,Or}Modifier

makortel committed:
Reduce copy-paste

makortel committed:
Add toReplaceWith() to _{And,Invert,Or}Modifier

Using the same bypass of contained-Modifier isChosen() status as with toModify()

makortel committed:
Replace isChosen() with boolean expression of Modifiers

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22210/3410

makortel commented:
@cmsbuild, please test workflow 250199.0,250199.17,250199.18,250202.0,250202.17,250202.18

Updates:
* Extended the `Modifier` boolean expressions with "not" (`~`) and "or" (`|`)
  * Latter is not strictly needed here, but @Dr15Jones preferred to review them both at the same time
* Added support for `toReplaceWith()` to "and", "or", and "not" for completeness (included in this PR in the spirit of previous comment)
* Replaced the `fastSim.isChosen()` with
```
modifyDigi_fastSimDigis = (fastSim & ~premix_stage1).makeProcessModifier(_fastSimDigis)
```

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #22210 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @lveldere, @emeschi, @civanch, @kpedro88, @fabozzi, @nsmith-, @rekovic, @cmsbuild, @mdhildreth, @thomreis, @slava77, @mommsen, @franzoni, @GurpreetSinghChahal, @fabiocos, @ssekmen, @davidlange6 can you please check and sign again.

Dr15Jones commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -1278,25 +1278,58 @@ def __call__(self,obj):
+      return
+    self._toModify(obj,func,**kw)
+  def _toModify(self,obj,func,**kw):
+    self._lhs._toModify(obj,func, **kw)
```
I assume you never call `self._rhs._*` because you assume the `_*` does the exact same work for `_lhs` and '_rhs`? That is because you ultimately expect that to call `Modifier._*`?

Would it be better to make `Modifier._*` a `@staticmethod` and all it directly?

makortel replied:
> I assume you never call self._rhs._* because you assume the _* does the exact same work for _lhs and 'rhs? That is because you ultimately expect that to callModifier.*`?

Yes, your assumption is correct.

> Would it be better to make Modifier._* a @staticmethod and all it directly?

I agree it would make the chain simpler to understand (and probably a bit faster as well since one doesn't have to trace the contained Modifier tree back to some leaf).

Dr15Jones commented on FWCore/ParameterSet/python/Config.py:
```diff
@@ -1370,6 +1413,10 @@ def isChosen(self):
     return _AndModifier(self,other)
+  def __invert__(self):
+    return _InvertModifier(self)
+  def __or__(self, other):
```
These allow `m1 & m2 & m3` to work, but I believe it would fail with `(m1 & m2) & m3` or maybe the failure is `m1 & (m2 & m3)`.  To make the later cases work, I think you need to add `__or__`, `__and__` and `__invert__` to `BoolModifierBase`.

makortel replied:
The `__or__`, `__and__` and `__invert__` are already part of `_BoolModifierBase`. I added couple of cases like  `(m1 & m2) & m3` to unit tests to actually test that.

cmsbuild commented:
-1

Tested at: ec9dc76ffa23032ec43708ac369375d745ece7bd

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26122/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26122/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.0_PREMIXUP15_PU25+PREMIXUP15_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.17_PREMIXUP17_PU25+PREMIXUP17_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.18_PREMIXUP18_PU25+PREMIXUP18_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250202.18_TTbar_13UP18+TTbar_13UP18+DIGIPRMXUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2465360
 - DQMHistoTests: Total failures: 28
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2465163
 - DQMHistoTests: Total skipped: 169
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.7099999999 KiB( 22 files compared)
 - Checked 111 log files, 9 edm output root files, 27 DQM output files

Dr15Jones commented:
The unit test failure is not from this pull request and always fails unless certain environment variables are set (which they are in all but these pull request tests).

makortel committed:
Add a couple of new cases to unit tests

makortel committed:
Declare Modifier._toModify() and _toReplacewith() as static

It's a bit simpler to avoid traversing the tree of Modifiers back to
some leaf.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-22210/3455

cmsbuild commented:
Pull request #22210 was updated. @perrotta, @smuzaffar, @prebello, @Dr15Jones, @lveldere, @emeschi, @civanch, @kpedro88, @fabozzi, @nsmith-, @rekovic, @cmsbuild, @mdhildreth, @thomreis, @slava77, @mommsen, @franzoni, @GurpreetSinghChahal, @fabiocos, @ssekmen, @davidlange6 can you please check and sign again.

makortel commented:
@cmsbuild, please test workflow 250199.0,250199.17,250199.18,250399.0,250202.0,250202.17,250202.18,250402.0

Update based on @Dr15Jones' comments
* Made `Modifier._toModify()` and `._toReplaceWith()` static to avoid traversing the tree of `Modifier`s
* Updated `Modifier` unit test to include cases like `(m1 & m2) & m3)`

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 225f34c959e20403e65410879dd7364d18f6bb1e

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26177/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testPythonParameterSet had ERRORS

cmsbuild commented:
Comparison job queued.

Dr15Jones commented:
+1
The unit test failure is expected for PR tests.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22210/26177/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.0_PREMIXUP15_PU25+PREMIXUP15_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.17_PREMIXUP17_PU25+PREMIXUP17_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250199.18_PREMIXUP18_PU25+PREMIXUP18_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250202.18_TTbar_13UP18+TTbar_13UP18+DIGIPRMXUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250399.0_FS_PREMIXUP15_PU25+FS_PREMIXUP15_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-22210/250402.0_TTbar_13+FS_TTbar_13_PRMXUP15_PU25+HARVESTUP15FS+MINIAODMCUP15FS

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 28
 - DQMHistoTests: Total histograms compared: 2498032
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2497855
 - DQMHistoTests: Total skipped: 176
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 1.03999999996 KiB( 23 files compared)
 - Checked 115 log files, 9 edm output root files, 28 DQM output files

kpedro88 commented:
+1
(enthusiastically!)

mommsen commented:
+1

perrotta commented:
+1
- Mostly Simulation
- It looks like it accomplishes what is mean to
- Moving premixing to modifiers, instead of customizations made in copy/pasted configs, also addresses #21704

thomreis commented:
+1

fabozzi commented:
+1

ssekmen commented:
+1

civanch commented:
+1

fabiocos commented:
+operations

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (but tests are reportedly failing). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1

fabiocos commented:
@Dr15Jones for my understanding, could you please clarify why is the failing unit test expected to do so?

makortel commented:
@fabiocos May I ask if you're waiting for @Dr15Jones' reply before merging? (I guess the fully signed + orp signature does not lead to automatic merge because of a failing unit test)

Dr15Jones commented:
testPythonParameterSet problem was discussed in this hypernews threads
https://hypernews.cern.ch/HyperNews/CMS/get/edmFramework/3398/1/1.html

essentially, if you run the tests in FWCore/PythonParameterSet but have not checked out FWCore/ParameterSet into your local are, this happens.

fabiocos commented:
@Dr15Jones ok, thanks for the clarification

fabiocos commented:
merge
